azure/services/ratelimiter.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services
+
+import (
+	"net/http"
+	"strconv"
+	"time"
+
+	"k8s.io/client-go/util/flowcontrol"
+)
+
+// RateLimitConfig indicates the rate limit config options.
+type RateLimitConfig struct {
+	// Enable rate limiting
+	AzureServiceRateLimit bool `json:"azureServiceRateLimit,omitempty" yaml:"azureServiceRateLimit,omitempty"`
+	// Rate limit QPS (Read)
+	AzureServiceRateLimitQPS float32 `json:"azureServiceRateLimitQPS,omitempty" yaml:"azureServiceRateLimitQPS,omitempty"`
+	// Rate limit Bucket Size
+	AzureServiceRateLimitBucket int `json:"azureServiceRateLimitBucket,omitempty" yaml:"azureServiceRateLimitBucket,omitempty"`
+	// Rate limit QPS (Write)
+	AzureServiceRateLimitQPSWrite float32 `json:"azureServiceRateLimitQPSWrite,omitempty" yaml:"azureServiceRateLimitQPSWrite,omitempty"`
+	// Rate limit Bucket Size
+	AzureServiceRateLimitBucketWrite int `json:"azureServiceRateLimitBucketWrite,omitempty" yaml:"azureServiceRateLimitBucketWrite,omitempty"`
+}
+
+type RateLimiter struct {
+	Reader flowcontrol.RateLimiter
+	Writer flowcontrol.RateLimiter
+}
+
+type RetryAfter struct {
+	Reader time.Time
+	Writer time.Time
+}
+
+const (
+	AzureServiceRateLimit            bool    = true
+	AzureServiceRateLimitQPS         float32 = 1.0
+	AzureServiceRateLimitBucket      int     = 1
+	AzureServiceRateLimitQPSWrite    float32 = 1.0
+	AzureServiceRateLimitBucketWrite int     = 1
+)
+
+// NewRateLimiter creates new read and write flowcontrol.RateLimiter from RateLimitConfig.
+func NewRateLimiter(config *RateLimitConfig) (flowcontrol.RateLimiter, flowcontrol.RateLimiter) {
```

Comment:
What do you think of moving this to a separate PR? it's a great improvement but I think for purposes of merging this quickly and getting it cherry-picked as a bug fix it would be beneficial to keep this PR as lean as possible. 

It would also be good to implement this across all services and doing a separate PR would also us to do that more easily.

Reply:
+1

Reply:
+1 to splitting the PR

Reply:
rate limiting will end up here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2572